### PR TITLE
Updated RTE with Redactor

### LINKED ISSUE #669: EE 6 RTE field enhancements
Two feature requests for the new RTE field in EE 6:

1. Add a "source" button to reveal the HTML source of the field contents to allow for custom HTML editing.

2. Add the ability to restrict the height of the field.  Otherwise, long text requires constant scrolling from mid/bottom back to top to be able to click the formatting buttons to apply formats.  It's virtually unusable like this.  Even with the height of the field set to 200 or 400, it does not limit to that height if the contents of the field are longer.

### PR DESCRIPTION
**Pull request by dougblackjr**
## Overview

Adds initial support for Redactor in the RTE field.

Resolves [#669](https://github.com/ExpressionEngine/ExpressionEngine/issues/669).

## Nature of This Change

- [ ] 🐛 Fixes a bug
- [X] 🚀 Implements a new feature
- [X] 🛁 Refactors existing code
- [X] 💅 Fixes coding style
- [ ] ✅ Adds tests
- [ ] 👽 Adds new dependency
- [ ] 🔥 Removes unused files / code
- [ ] 🔒 Improves security

## Is this backwards compatible?

- [X] Yes
- [ ] No

### CODE REVIEW

**Comment by litzinger** [comment]
I just checked out this PR, ran the updates to RTE 2.1.0, created a new field, added it to a channel, and then tried to create a new entry in that channel and received this error:

```
Object of class stdClass could not be converted to string

/var/www/html/ExpressionEngine/system/ee/legacy/libraries/Javascript.php:659

Stack Trace: Please include when reporting this error

#0 ExpressionEngine/system/ee/legacy/libraries/Javascript.php(659): array_unique(Array, 2)
#1 ExpressionEngine/system/ee/legacy/libraries/Javascript.php(636): EE_Javascript->set_global('Rte.configs.def...', Array)
#2 ExpressionEngine/system/ee/ExpressionEngine/Addons/Rte/Service/CkeditorService.php(201): EE_Javascript->set_global(Array)
#3 ExpressionEngine/system/ee/ExpressionEngine/Addons/Rte/Service/CkeditorService.php(27): ExpressionEngine\Addons\Rte\Service\CkeditorService->insertConfigJsById()
#4 ExpressionEngine/system/ee/ExpressionEngine/Addons/rte/ft.rte.php(171): ExpressionEngine\Addons\Rte\Service\CkeditorService->init(Array, Object(ExpressionEngine\Addons\Rte\Model\Toolset))
#5 ExpressionEngine/system/ee/legacy/fieldtypes/EE_Fieldtype.php(320): Rte_ft->display_field(NULL)
#6 ExpressionEngine/system/ee/legacy/libraries/api/Api_channel_fields.php(377): EE_Fieldtype->display_publish_field(NULL)
#7 ExpressionEngine/system/ee/ExpressionEngine/Model/Content/FieldFacade.php(250): Api_channel_fields->apply('display_publish...', Array)
#8 ExpressionEngine/system/ee/ExpressionEngine/Model/Content/Display/FieldDisplay.php(70): ExpressionEngine\Model\Content\FieldFacade->getForm()
#9 ExpressionEngine/system/ee/ExpressionEngine/View/publish/partials/publish_form.php(83): ExpressionEngine\Model\Content\Display\FieldDisplay->getForm()
```

**Comment by dougblackjr** [comment]
@litzinger I am unable to reproduce this on EE6.0.3. I did push a change to cast that section as an array. Can you tell me what version of PHP and EE you are running? Is this a new install or an existing site?

**Comment by litzinger** [comment]
6.0.3, but I checked out this commit specifically. PHP 7.4.13

**Comment by obfuscode** [comment]
@dougblackjr Looks like some of the tests are still failing for this - if you're able to take a look to see if you can get those to pass or loop @intoeetive into it at some point.

**Comment by dougblackjr** [comment]
@obfuscode Working on this now. Part of this requires and updated add-on which adds a database field; is there a recommended way to do this in the testing suite so that it passes in the PR?

cc: @intoeetive

**Comment by intoeetive** [comment]
When test are set up, the DB dump is loaded from `support/sql/database6.0.0.sql`, so if core DB structure is changes, that needs to be reflected.

If however add-on is installable, it's best to have installation/uninstallation as part of testing siute (I believe you can reuse some code from add-ons tests folder)

File: system/ee/ExpressionEngine/Addons/rte/ft.rte.php [hunk 1]
```diff
@@ -214,6 +221,7 @@ public function display_field($data)
      */
     public function grid_display_field($data)
     {
+        // @TODO: Repeat this code
         RteHelper::includeFieldResources();
         $configHandle = RteHelper::insertConfigJsById(!empty($this->settings['toolset_id']) ? $this->settings['toolset_id'] : null);
```
Comment (intoeetive) [inline system/ee/ExpressionEngine/Addons/rte/ft.rte.php:229]: @dougblackjr does this need to be updated to support Redactor as well?

File: system/ee/ExpressionEngine/Addons/rte/views/redactor-toolbar.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            <?php foreach ($buttons as $btnHandle => $btn) :?>
+                <p>
+                    <label><?=$btn['label']?></label>
+                    <?=form_checkbox('settings[toolbar][buttons][]', $btnHandle, in_array($btnHandle, $settings['toolbar']['buttons']))?>
+                </p>
+            <?php endforeach;?>
+        </div>
+        
+        <div class="redactor-toolbar">
+            <h2>Plugins</h2>
```
Comment (intoeetive) [inline system/ee/ExpressionEngine/Addons/rte/views/redactor-toolbar.php]: The H2 headings (just as any text) needs to come from lang files

**Review by intoeetive — CHANGES_REQUESTED** [review]
@dougblackjr please see the comments I've made.

Additionally, since RTE is installed with EE, we need to make sure the new column is added when they update to 6.1.0. So it needs to be updateable both as module and part of EE default install.

As of the tests, since the only DB change is adding `toolset_type` column, you should be fine just adding it to DB dump. From there, is the tests fail, you should look at artifacts in GitHub actions for screenshots, or run the failed tests locally

File: system/ee/ExpressionEngine/Addons/rte/ft.rte.php (hunk 1 above)
Comment (dougblackjr) [reply system/ee/ExpressionEngine/Addons/rte/ft.rte.php:229]: @intoeetive No, with the changes to the helper, it should cover any providers.

File: system/ee/ExpressionEngine/Addons/rte/views/redactor-toolbar.php (hunk 2 above)
Comment (dougblackjr) [reply system/ee/ExpressionEngine/Addons/rte/views/redactor-toolbar.php]: Thank you, fixed in latest commit.

**Review by obfuscode — DISMISSED** [review]

**Comment by obfuscode** [comment]
@dougblackjr Can you resolve these last conflicts and then I think we're good to merge.

**Review by intoeetive — CHANGES_REQUESTED** [review]
The tests are failing for Fluid field and for Updater. @dougblackjr more details in your Slack.

Also I think it's worth to manually check how the field behaves as Grid column and Low variable - that is not covered with automated tests.

**Comment by intoeetive** [comment]
The tests are now passing, however this is not ready to be merged:
- the UI/UX of Redactor configuration needs improvement
- the buttons selected for Redactor do not seem to be respected
- Cypress tests need to be written

**Pull request closed without merging**